### ISSUE TITLE
Support for passing basic auth credentials from Python client

### ISSUE THREAD

**Issue by kesavkolla** [apache/pulsar#2499]
If broker is setup with BasicAuth (**AuthenticationProviderBasic**) currently Python client doesn't have an ability to pass the credentials.  Please add support in Python for sending basic auth credentials while connecting to topic.

**Comment by sijie** [comment]
I think this is a documentation issue to documentation how to pass basic authentication credentials from Python client.

**Comment by codelipenghui** [comment]
@tuteng Could you help check if the feature is supported?

**Comment by 0x5e** [comment]
+1
I have no idea how to extend auth in python. There's only java documentation.
https://pulsar.apache.org/docs/en/next/security-extending/

**Comment by tuteng** [comment]
I think the current python client does not support this feature, the python client is a wrapper around cpp, according to the code here https://github.com/apache/pulsar/tree/master/pulsar-client-cpp/lib/auth, the cpp client does not implement basic authentication @sijie @codelipenghui @0x5e @kesavkolla I'll mark it as `help-wanted` first, and see if anyone wants to contribute this feature

**Comment by 0x5e** [comment]
Thanks @tuteng , I have tried to build a c++ auth plugin and finally succeed. Not basic auth, just custom auth for my company, they are similar.
If anyone interested I can provide some code when my test completed.

**Comment by HenriqueFSilva** [comment]
> Thanks @tuteng , I have tried to build a c++ auth plugin and finally succeed. Not basic auth, just custom auth for my company, they are similar. If anyone interested I can provide some code when my test completed.

Hi @0x5e , can you provide the code for the custom c++ authentication plugin you implemented? Thanks in advance

**Comment by 0x5e** [comment]
AuthCustom.h
```c++
#pragma once

#include <string>
#include <map>

#if __has_include(<pulsar/Authentication.h>)
    #include <pulsar/Authentication.h>
#else
namespace pulsar {

    class AuthenticationDataProvider;
    class Authentication;

    typedef std::shared_ptr<AuthenticationDataProvider> AuthenticationDataPtr;
    typedef std::shared_ptr<Authentication> AuthenticationPtr;
    typedef std::map<std::string, std::string> ParamMap;

    enum Result
    {
        ResultOk,  /// Operation successful
    };

    class AuthenticationDataProvider {
        public:
            virtual ~AuthenticationDataProvider();
            virtual bool hasDataForTls();
            virtual std::string getTlsCertificates();
            virtual std::string getTlsPrivateKey();
            virtual bool hasDataForHttp();
            virtual std::string getHttpAuthType();
            virtual std::string getHttpHeaders();
            virtual bool hasDataFromCommand();
            virtual std::string getCommandData();

        protected:
            AuthenticationDataProvider();
    };

    class Authentication {
        public:
            virtual ~Authentication();
            virtual const std::string getAuthMethodName() const = 0;
            virtual Result getAuthData(AuthenticationDataPtr& authDataContent) {
                authDataContent = authData_;
                return ResultOk;
            }
            static ParamMap parseDefaultFormatAuthParams(const std::string& authParamsString);

        protected:
            Authentication();
            AuthenticationDataPtr authData_;
            friend class ClientConfiguration;
    };

}
#endif


namespace pulsar {

    AuthenticationDataProvider::AuthenticationDataProvider() {}

    AuthenticationDataProvider::~AuthenticationDataProvider() {}

    bool AuthenticationDataProvider::hasDataForTls() { return false; }

    std::string AuthenticationDataProvider::getTlsCertificates() { return "none"; }

    std::string AuthenticationDataProvider::getTlsPrivateKey() { return "none"; }

    bool AuthenticationDataProvider::hasDataForHttp() { return false; }

    std::string AuthenticationDataProvider::getHttpAuthType() { return "none"; }

    std::string AuthenticationDataProvider::getHttpHeaders() { return "none"; }

    bool AuthenticationDataProvider::hasDataFromCommand() { return false; }

    std::string AuthenticationDataProvider::getCommandData() { return "none"; }

    Authentication::Authentication() {}

    Authentication::~Authentication() {}

// https://github.com/apache/pulsar/blob/master/pulsar-client-cpp/lib/auth/AuthToken.h
class AuthDataCustom : public AuthenticationDataProvider {
    public:
        AuthDataCustom(const std::string &commandData);
        ~AuthDataCustom();

        bool hasDataForHttp();
        // std::string getHttpHeaders();
        bool hasDataFromCommand();
        std::string getCommandData();

    private:
        const std::string commandData_;
};

// https://github.com/apache/pulsar/blob/master/pulsar-client-cpp/include/pulsar/Authentication.h
class AuthCustom : public Authentication  {
    public:
        AuthCustom(AuthenticationDataPtr&);
        ~AuthCustom();

        static AuthenticationPtr create(const std::string& authParamsString);

        const std::string getAuthMethodName() const;
        Result getAuthData(AuthenticationDataPtr& authDataToken);

    private:
        AuthenticationDataPtr AuthDataCustom_;
};

}  // namespace pulsar

```

AuthCustom.cc
```c++
#include "AuthCustom.h"

namespace pulsar {

// AuthDataCustom

// https://github.com/apache/pulsar/blob/master/pulsar-client-cpp/lib/auth/AuthToken.cc
AuthDataCustom::AuthDataCustom(const std::string &commandData) : commandData_(commandData) {}

AuthDataCustom::~AuthDataCustom() {}

bool AuthDataCustom::hasDataForHttp() { return false; }

bool AuthDataCustom::hasDataFromCommand() { return true; }

std::string AuthDataCustom::getCommandData() { return commandData_; }

// AuthCustom

AuthCustom::AuthCustom(AuthenticationDataPtr &AuthDataCustom) : AuthDataCustom_(AuthDataCustom) {}

AuthCustom::~AuthCustom() {}

AuthenticationPtr AuthCustom::create(const std::string &commandData) {
    AuthenticationDataPtr AuthDataCustom = AuthenticationDataPtr(new AuthDataCustom(commandData));
    return AuthenticationPtr(new AuthCustom(AuthDataCustom));
}

const std::string AuthCustom::getAuthMethodName() const { return "auth1"; }

Result AuthCustom::getAuthData(AuthenticationDataPtr &authDataContent) {
    authDataContent = AuthDataCustom_;
    return ResultOk;
}


extern "C" Authentication* create(const std::string &commandData) {
    AuthenticationDataPtr AuthDataCustom = AuthenticationDataPtr(new AuthDataCustom(commandData));
    return (new AuthCustom(AuthDataCustom));
}

}  // namespace pulsar

```


@HenriqueFSilva this is the code. I forget the detail and not work in that company now. Hope it can help you:)

**Comment by momo-jun** [comment]
Confirmed by @nodece - this issue has a dependency on #15822 and needs to support basic auth on node.js + python client.

**Comment by nodece** [comment]
Link to #17482.

**Comment by nodece** [comment]
Closed by #17482.
